File: binary-search-tree/bst.js
// using ES6 class syntax 
class Node {
	constructor(value, left = null, right = null) {
		this.value = value;
		this.left = left;
		this.right = right;
	}
}

class MyBinarySearchTree {
	constructor(root = null) {
		this.root = root;
	}

	// bst sort
	inorderTraversal(root = this.root) {
		if (root === null) return;
		this.inorderTraversal(root.left);
		console.log(root.value);
		this.inorderTraversal(root.right);
	}

	// searching a node of a given value
	search(target, root = this.root) {
		if (root === null) {
			console.log(false);
			return;
		}
		if (root.value === target) console.log(true);
		else if (root.value < target) this.search(target, root.right);
		else this.search(target, root.left);
	}

	// searching a node of a given value non-recursive version 
	searchNR(target, root = this.root) {
		while (root !== null) {
			if (target === root.value) {
				console.log(true);
				return;
			} else if (target > root.value) {
				root = root.right;
			} else {
				root = root.left;
			}
		}
		console.log(false);
		return;
	}

	// insert a node to a bst
	insert(value) {
		const node = new Node(value);
		if (this.root === null) {
			this.root = node;
			return;
		}

		p = this.root;
		while ((p.left !== null ) || (p.right !== null)) {
			if ((p.left === null) && (value < p.value)) {
				p.left = node;
				break;
			} else if ((p.right === null) && (value >= p.value)) {
				p.right = node;
				break;
			} else if (value >= p.value) p = p.right;
			  else if (value < p.value) p = p.left;
		}
		if (value >= p.value) p.right = node;
		else p.left = node;
	}

	// delete a node from a bst
	delete(value) {
		if (this.root === null) return false;
		let p = this.root;
		let prev = this.root;
		while (p !== null) {
			if (p.value > value) {
				prev = p;
				p = p.left
			} else if (p.value < value) {
				prev = p;
				p = p.right;
			} else break;
		}
		if (p === null) return false;
		if (p.left === null && p.right === null) {
			if (prev.left === p) prev.left = null;
			else if (prev.right === p) prev.right = null;
			return true;
		}
		if (p.left !== null || p.right !== null) {
			if (p.left !== null) {
				if (prev.left === p) prev.left = p.left;
				else if (prev.right === p) prev.right = p.left;
			} else if (p.right !== null) {
				if (prev.left === p) prev.left = p.right;
				else if (prev.right === p) prev.right = p.right;
			}
			return true;
		}
	}
}

File: binary-search-tree/bst.py
from typing import Any

class Node:
	def __init__(self, value: Any, left: 'Node'= None, right: 'Node'=None) -> None:
		self.value: Any = value
		self.left: 'Node' = left
		self.right: 'Node' = right 


class MyBinarySearchTree:
	'''
	implement binary search tree using python
	'''
	def __init__(self, root: 'Node'=None) -> None:
		self.root = root

	def inorder_traversal(self, root: 'Node') -> None: 
		'''
		bst sort
		'''
		if root == None:
			return
		self.inorder_traversal(root.left)
		print(root.value)
		self.inorder_traversal(root.right)

	def search(self, target: Any, root: 'Node'=None) -> None:
		'''
		search a node of a given value
		'''
		if root == None:
			print(False)
			return
		if root.value == target:
			print(True)
		elif root.value < target:
			self.search(target, root.right)
		else:
			self.search(target, root.left)

	def search_n_r(self, target: Any, root: 'Node'=None) -> bool:
		'''
		non-recursive version of search
		'''
		while root != None:
			if target == root.value:
				return True
			elif target > root.value:
				root = root.right
			else:
				root = root.left
		return False

	def insert(self, value: Any) -> None:
		'''
		insert a node to bst
		'''
		node = Node(value)
		if self.root == None:
			self.root = node 
			return

		p = self.root
		while p.left != None or p.right != None:
			if p.left == None and value < p.value:
				p.left = node 
				break
			elif p.right == None and value >= p.value:
				p.right = node 
				break 
			elif value >= p.value:
				p = p.right
			elif value < p.value:
				p = p.left
		if value >= p.value:
			p.right = node 
		else:
			p.left = node

	def delete(self, value: Any) -> None:
		'''
		delete a node from bst
		'''
		if self.root == None:
			return False
		p = self.root 
		prev = self.root
		while p != None:
			if p.value > value:
				prev = p
				p = p.left
			elif p.value < value:
				prev = p
				p = p.right
			else:
				break
		if p == None:
			return False
		if p.left == None and p.right == None:
			if prev.left == p:
				prev.left = None
			elif prev.right == p:
				prev.right = None
			return True
		if p.left != None or p.right != None:
			if p.left != None:
				if prev.left == p:
					prev.left = p.left
				elif prev.right == p:
					prev.right = p.left
			elif p.right != None:
				if prev.left == p:
					prev.left = p.right
				elif prev.right == p:
					prev.right = p.right
			return True
